ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PullQueryMetricsFunctionalTest.java

```diff
@@ -0,0 +1,283 @@
+  private static final String AN_AGG_KEY = "USER_1";
+  private static final String A_STREAM_KEY = "PAGE_1";
+
+  private static final PhysicalSchema AGGREGATE_SCHEMA = PhysicalSchema.from(
+      LogicalSchema.builder()
+          .keyColumn(ColumnName.of("USERID"), SqlTypes.STRING)
+          .valueColumn(ColumnName.of("COUNT"), SqlTypes.BIGINT)
+          .build(),
+      SerdeFeatures.of(),
+      SerdeFeatures.of()
+  );
+
+  private static final ImmutableMap<String, String> TABLE_TAGS = ImmutableMap.of(
+      "ksql_service_id", "default_",
+      "query_source", "non_windowed",
+      "query_plan_type", "key_lookup",
+      "query_routing_type", "source_node"
+  );
+
+  private static final ImmutableMap<String, String> STREAMS_TAGS = ImmutableMap.of(
+      "ksql_service_id", "default_",
+      "query_source", "non_windowed_stream",
+      "query_plan_type", "unknown",
+      "query_routing_type", "source_node"
+  );
+
+  private static final IntegrationTestHarness TEST_HARNESS = IntegrationTestHarness.build();
+
+  private static final TestKsqlRestApp REST_APP = TestKsqlRestApp
+      .builder(TEST_HARNESS::kafkaBootstrapServers)
+      .withProperty(KsqlConfig.METRIC_REPORTER_CLASSES_CONFIG, TestMetricsReporter.class.getName())
+      .withProperty(KsqlConfig.KSQL_QUERY_STREAM_PULL_QUERY_ENABLED, true)
+      .build();
+
+  @ClassRule
+  public static final RuleChain CHAIN = RuleChain.outerRule(TEST_HARNESS).around(REST_APP);
+
+  @Rule
+  public final Timeout timeout = Timeout.seconds(60);
+
+  @BeforeClass
+  public static void setUpClass() {
+    TEST_HARNESS.ensureTopics(PAGE_VIEW_TOPIC);
+
+    TEST_HARNESS.produceRows(PAGE_VIEW_TOPIC, PAGE_VIEWS_PROVIDER, FormatFactory.KAFKA, FormatFactory.JSON);
+
+    RestIntegrationTestUtil.createStream(REST_APP, PAGE_VIEWS_PROVIDER);
+
+    RestIntegrationTestUtil.makeKsqlRequest(REST_APP, "CREATE TABLE " + AGG_TABLE + " AS "
+        + "SELECT USERID, COUNT(1) AS COUNT FROM " + PAGE_VIEW_STREAM + " GROUP BY USERID;"
+    );
+
+    waitForTableRows();
+  }
+
+  @Before
+  public void setUp() {
+  }
+
+  @After
+  public void tearDown() {
+  }
+
+  @AfterClass
+  public static void classTearDown() {
+  }
+
+  public static class TestMetricsReporter implements MetricsReporter {
+    static final ConcurrentHashMap<MetricName, KafkaMetric> METRICS = new ConcurrentHashMap<>();
+
+    @Override
+    public void init(final List<KafkaMetric> metrics) {
+      for (final KafkaMetric metric : metrics) {
+        METRICS.put(metric.metricName(), metric);
+      }
+    }
+
+    @Override
+    public void metricChange(final KafkaMetric metric) {
+      METRICS.put(metric.metricName(), metric);
+    }
+
+    @Override
+    public void metricRemoval(final KafkaMetric metric) {
+      METRICS.remove(metric.metricName());
+    }
+
+    @Override
+    public void close() {
+    }
+
+    @Override
+    public void configure(final Map<String, ?> configs) {
+    }
+  }
+
+  @Test
```

Comment:
Unfortunately, I have to add all assertions in one test since I haven't managed to find a way for the metrics to get cleaned between tests. I tried even creating a new rest app for each test but even that didn't work.

Reply:
I think this is fine for now. A test is better than no test!

We can always iterate on the metric testing strategy in the future.